[cloudevents] Add override example (#3543)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/pubsub/pubsub-cloudevents.md

- Content-type for proper deserialization of event data
- Verification of sender application

You can choose any of the three methods for publish a CloudEvent via pub/sub:

> 1. Write a pub/sub event, which is then automatically wrapped by Dapr in a CloudEvent envelope.

Reviewer: ```suggestion
1. Send a pub/sub event, which is then wrapped by Dapr in a CloudEvent envelope.
```

> 1. Retain the CloudEvents attributes automatically provided by Dapr while overriding any of the standard CloudEvent properties.

Reviewer: ```suggestion
1. Replace specific CloudEvents attributes provided by Dapr by overriding the standard CloudEvent properties.
```

> 1. Supply your own full CloudEvent envelope.

Reviewer: ```suggestion
1. Write your own CloudEvent envelope as part of the pub/sub event.
```

## Dapr-generated CloudEvents example

> Writing a publish operation to Dapr automatically wraps it in a CloudEvent envelope containing the following fields:

Reviewer: ```suggestion
Sending a publish operation to Dapr automatically wraps it in a CloudEvent envelope containing the following fields:
```

  "tracestate": "",
  "data" : "<note><to></to><from>user2</from><message>Order</message></note>",
  "id" : "id-1234-5678-9101",
  "specversion" : "1.0",
  "datacontenttype" : "text/xml",
  "subject" : "Test XML Message",
  "source" : "https://example.com/message",
  "type" : "xml.message",
   "time" : "2020-09-23T06:23:21Z"
}
```

## Replace Dapr generated CloudEvents values

Dapr automatically generates several CloudEvent properties. You can replace these generated CloudEvent properties by providing the following optional metadata key/value:

- `cloudevent-id`: overrides `id`
- `cloudevent-source`: overrides `source`
- `cloudevent-type`: overrides `type`
- `cloudevent-traceid`: overrides `traceid`
- `cloudevent-tracestate`: overrides `tracestate`
- `cloudevent-traceparent`: overrides `traceparent`

The ability to replace CloudEvents properties using these metadata properties applies to all pub/sub components.

### Example

For example, to replace the `source` and `id` values from [the CloudEvent example above]({{< ref "#cloudevents-example" >}}) in code:

{{< tabs "Python" ".NET" >}}
 <!-- Python -->
{{% codetab %}}

```python

> with DaprClient() as client:

Reviewer: Need to remove the for loop entirely. A loop of one does not do anything. See C# example below.

    for i in range(1):
        order = {'orderId': i}
        # Publish an event/message using Dapr PubSub
        result = client.publish_event(
            pubsub_name='order_pub_sub',
            topic_name='orders',
            publish_metadata={'cloudevent-id: 'd99b228f-6c73-4e78-8c4d-3f80a043d317', cloudevent-source: 'payment'}
        )
```

{{% /codetab %}}

 <!-- .NET -->
{{% codetab %}}

```csharp

> for (int i = 1) {

Reviewer: I meant removing the for loop entirely. Like this

```
var order = new Order("order1");
using var client = new DaprClientBuilder().Build();

// Override cloudevent metadata
var metadata = new Dictionary<string,string>() {
       { "cloudevent.source", "payment" },
       { "cloudevent.id", "d99b228f-6c73-4e78-8c4d-3f80a043d317" }
}

// Publish an event/message using Dapr PubSub
await client.PublishEventAsync("order_pub_sub", "orders", order, metadata);
Console.WriteLine("Published data: " + order);

await Task.Delay(TimeSpan.FromSeconds(1));
```

Author: ohh ok I misunderstood - fixing.

}
```

### Override CloudEvent values

> You can override CloudEvent values by replacing some metadata with the following:

Reviewer 2: ```suggestion
When not provided Dapr automatically generates several Cloud Event properties. You can override these generated CloudEvent properties by providing the following optional metadata keys:
```

I would add that this overrides the values that would otherwise be automatically generated by Dapr.

Also as an aside, this isn't the main way to pass in a trace ID / trace state.

- `cloudevent-id`: overrides `id`
- `cloudevent-source`: overrides `source`
- `cloudevent-type`: overrides `type`
- `cloudevent-traceid`: overrides `traceid`
- `cloudevent-tracestate`: overrides `tracestate`
- `cloudevent-traceparent`: overrides `traceparent`

{{% alert title="Note" color="primary" %}}
The following metadata can't be overridden:
- `data`
- `topic`
- `pubsub`
- `datacontenttype`
{{% /alert %}}

For example, to override the applicable values from [the first CloudEvent example above]({{< ref "#cloudevents-example" >}}):

```json
{
  "topic": "orders",
  "pubsubname": "order_pub_sub",

>   "cloudevent-traceid": "00-113ad9c4e42b27583ae98ba698d54255-e3743e35ff56f219-01",

Reviewer 2: Because this isn't the recommended way to specify the TraceID and TraceState I would drop these from the example.

  "cloudevent-tracestate": "",
  "data": {
    "orderId": 1
  },

>   "cloudevent-id": "5929aaac-a5e2-4ca1-859c-edfe73f11565",

Reviewer 2: Specifying this in the payload is actually incorrect. It needs to be sent in the Publish Metadata request.

For example in Python:

```python
    resp = d.publish_event(
        pubsub_name='pubsub',
        topic_name='TOPIC_A',
        data=json.dumps(req_data),
        publish_metadata={'cloudevent-id': '5929aaac-a5e2-4ca1-859c-edfe73f11565'}
    )
```

}
```

> ## Override Dapr-generated CloudEvents values

Reviewer: ```suggestion
## Replace Dapr generated CloudEvents values
```

> Dapr automatically generates several CloudEvent properties. You can override these generated CloudEvent properties by providing the following optional metadata keys:

Reviewer: ```suggestion
Dapr automatically generates several CloudEvent properties. You can replace these generated CloudEvent properties by providing the following optional metadata key/value:
```

- `cloudevent-id`: overrides `id`
- `cloudevent-source`: overrides `source`
- `cloudevent-type`: overrides `type`
- `cloudevent-traceid`: overrides `traceid`
- `cloudevent-tracestate`: overrides `tracestate`
- `cloudevent-traceparent`: overrides `traceparent`

> The CloudEvent override metadata properties apply globally for all pub/sub components.

Reviewer: ```suggestion
The ability to replace CloudEvents properties using these metadata properties applies to all pub/sub components.
```

### Example

> To override the `source` and `id` values from [the CloudEvent example above]({{< ref "#cloudevents-example" >}}) in code:

Reviewer: ```suggestion
For example, to replace the `source` and `id` values from [the CloudEvent example above]({{< ref "#cloudevents-example" >}}) in code:
```

{{< tabs "Python" ".NET" >}}
 <!-- Python -->
{{% codetab %}}

```python
with DaprClient() as client:

>     for i in range(1, 10):

Reviewer: We can keep this simple and not show how to do this 10 times. I would change the code to a single call.

        order = {'orderId': i}
        # Publish an event/message using Dapr PubSub
        result = client.publish_event(
            pubsub_name='order_pub_sub',
            topic_name='orders',
            publish_metadata={'cloudevent-id: 'd99b228f-6c73-4e78-8c4d-3f80a043d317', cloudevent-source: 'payment'}
        )
```

{{% /codetab %}}

 <!-- .NET -->
{{% codetab %}}

```csharp

> for (int i = 1; i <= 10; i++) {

Reviewer: Same here. Single call.

    var order = new Order(i);
    using var client = new DaprClientBuilder().Build();

   // Override cloudevent metadata
    var metadata = new Dictionary<string,string>() {
        { "cloudevent.source", "payment" },
        { "cloudevent.id", "d99b228f-6c73-4e78-8c4d-3f80a043d317" }
    }

    // Publish an event/message using Dapr PubSub
    await client.PublishEventAsync("order_pub_sub", "orders", order, metadata);
    Console.WriteLine("Published data: " + order);

    await Task.Delay(TimeSpan.FromSeconds(1));
}
```

{{% /codetab %}}

{{< /tabs >}}

The JSON payload then reflects the new `source` and `id` values:

```json
{
  "topic": "orders",
  "pubsubname": "order_pub_sub",
  "traceid": "00-113ad9c4e42b27583ae98ba698d54255-e3743e35ff56f219-01",
  "tracestate": "",
  "data": {
    "orderId": 1
  },
  "id": "d99b228f-6c73-4e78-8c4d-3f80a043d317",
  "specversion": "1.0",
  "datacontenttype": "application/json; charset=utf-8",
  "source": "payment",
  "type": "com.dapr.event.sent",
  "time": "2020-09-23T06:23:21Z",
  "traceparent": "00-113ad9c4e42b27583ae98ba698d54255-e3743e35ff56f219-01"
}
```

{{% alert title="Important" color="warning" %}}

> While you can override `traceid`/`traceparent` and `tracestate`, doing this may interfere with tracing tools. It's recommended to use Open Telementry and Zipkin protocols for distributed traces. [Learn more about distributed tracing.]({{< ref tracing-overview.md >}}) 

Reviewer: ```suggestion
While you can replace `traceid`/`traceparent` and `tracestate`, doing this may interfere with tracing events and report inconsistent result in tracing tools. It's recommended to use Open Telementry for distributed traces. [Learn more about distributed tracing.]({{< ref tracing-overview.md >}}) 
```
